contracts/pcv/balancer/BalancerPCVDepositWeightedPool.sol

```diff
@@ -0,0 +1,323 @@
+
+        tokenOracles = _tokenOracles;
+
+        // set cached values for token addresses & indexes
+        bool tokenFound = false;
+        address _fei = address(fei());
+        for (uint256 i = 0; i < poolAssets.length; i++) {
+            tokenOraclesMapping[IERC20(address(poolAssets[i]))] = _tokenOracles[i];
+            if (address(poolAssets[i]) == _token) {
+                tokenFound = true;
+                tokenIndexInPool = uint8(i);
+                token = IERC20(address(poolAssets[i]));
+            }
+            if (address(poolAssets[i]) == _fei) {
+                feiInPool = true;
+                feiIndexInPool = uint8(i);
+            }
+        }
+        // check that the token is in the pool
+        require(tokenFound, "BalancerPCVDepositWeightedPool: token not in pool.");
+
+        // check that token used for account is not FEI
+        require(_token != _fei, "BalancerPCVDepositWeightedPool: token must not be FEI.");
+    }
+
+    /// @notice sets the oracle for a token in this deposit
+    function setOracle(address _token, address _newOracle) external onlyGovernorOrAdmin {
+        // we must set the oracle for an asset that is in the pool
+        address oldOracle = address(tokenOraclesMapping[IERC20(_token)]);
+        require(oldOracle != address(0), "BalancerPCVDepositWeightedPool: invalid token");
+
+        // set oracle in the map
+        tokenOraclesMapping[IERC20(_token)] = IOracle(_newOracle);
+
+        // emit event
+        emit OracleUpdate(
+            msg.sender,
+            _token,
+            oldOracle,
+            _newOracle
+        );
+    }
+
+    /// @notice returns total balance of PCV in the Deposit, expressed in "token"
+    function balance() public view override returns (uint256) {
+        // read oracle values
+        uint256[] memory underlyingPrices = _readOracles();
+
+        // get BPT token price
+        uint256 bptPrice = _getBPTPrice(underlyingPrices);
+
+        // compute balance in USD value
+        uint256 bptBalance = IWeightedPool(poolAddress).balanceOf(address(this));
+        Decimal.D256 memory bptValueUSD = Decimal.from(bptBalance).mul(bptPrice).div(1e18);
+
+        // return balance in "token" value
+        uint256 _balanceInToken = bptValueUSD.mul(1e18).div(underlyingPrices[tokenIndexInPool]).asUint256();
+        // if FEI is in the pair, return only the value of asset, and does not
+        // count the protocol-owned FEI in the balance. For instance, if the pool
+        // is 80% WETH and 20% FEI, balance() will return 80% of the USD value
+        // of the balancer pool tokens held by the contract, denominated in
+        // "token" (and not in USD).
+        if (feiInPool) {
+            uint256[] memory _weights = IWeightedPool(poolAddress).getNormalizedWeights();
+            _balanceInToken = _balanceInToken * _weights[tokenIndexInPool] / 1e18;
+        }
+        return _balanceInToken;
```

Comment:
This looks more like resistantBalance. I'd use the vault balances here I think. Is there a reason you didn't?

Reply:
ah yes, I'll change implementation... At first I didn't want to implement resistantBalanceAndFei, and use the default behavior (that returns (balance(), 0), but then realized there could be FEI and had to reimplement a proper method.

I'll read the pool balances directly for the normal balance() call